### PR TITLE
authz: add "Schedule now" button to permissions pages

### LINKED ISSUE #8990: Admin button for refreshing permissions for a user / repo
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/web/src/repo/settings/RepoSettingsMirrorPage.tsx#L303:40

### PR DESCRIPTION
**Pull request by unknwon**
Adds "Schedule now" button to permissions pages of both user and repository settings, which basically hits corresponding GraphQL mutations.

image

This finally fixes #8990 as the last mile.

### CODE REVIEW

**Comment by unknwon** [comment]
cc @christinaforney do you think the screenshot in the description is good enough? (both user and repository ones are look very similar in layout, so just provide one of them is sufficient IMO)

File: web/src/user/settings/auth/UserSettingsPermissionsPage.tsx [hunk 1]
```diff
@@ -24,29 +30,58 @@ export const UserSettingsPermissionsPage: React.FunctionComponent<{ user: GQL.IU
...
+                            <tr>
+                                <th>Last incremental sync</th>
+                                <td>
+                                    <Timestamp date={user.permissionsInfo.updatedAt} />
+                                </td>
+                                <td className="text-muted">Updated by repository permissions syncing</td>
+                            </tr>
+                        </tbody>
+                    </table>
+                    <ScheduleUserPermissionsSyncActionContainer user={user} history={history} />
```
Comment (eseliger) [inline web/src/user/settings/auth/UserSettingsPermissionsPage.tsx:56]: it seems like this should be hidden when no site-admin?

File: web/src/site-admin/backend.tsx [hunk 2]
```diff
@@ -208,6 +208,40 @@ export function checkMirrorRepositoryConnection(
...
+                scheduleRepositoryPermissionsSync(repository: $repository) {
+                    alwaysNil
+                }
+            }
+        `,
+        args
+    ).pipe(
+        map(dataOrThrowErrors),
+        tap(() => resetAllMemoizationCaches()),
+        map(() => undefined)
```
Comment (eseliger) [inline web/src/site-admin/backend.tsx:224]:
```suggestion
        mapTo(undefined)
```

File: web/src/site-admin/backend.tsx [hunk 3]
```diff
@@ -208,6 +208,40 @@ export function checkMirrorRepositoryConnection(
...
+                scheduleUserPermissionsSync(user: $user) {
+                    alwaysNil
+                }
+            }
+        `,
+        args
+    ).pipe(
+        map(dataOrThrowErrors),
+        tap(() => resetAllMemoizationCaches()),
+        map(() => undefined)
```
Comment (eseliger) [inline web/src/site-admin/backend.tsx:241]:
```suggestion
        mapTo(undefined)
```

File: web/src/repo/settings/RepoSettingsPermissionsPage.tsx [hunk 4]
```diff
@@ -24,29 +30,60 @@ export const RepoSettingsPermissionsPage: React.FunctionComponent<{ repo: GQL.IR
...
+
+class ScheduleRepositoryPermissionsSyncActionContainer extends React.PureComponent<
+    ScheduleRepositoryPermissionsSyncActionContainerProps
+> {
+    public render(): JSX.Element | null {
+        return (
+            <ActionContainer
+                title="Manually schedule a permissions sync"
+                description={
+                    <div>Site admins are able to manually schedule a permissions sync for this repository.</div>
```
Comment (eseliger) [inline web/src/repo/settings/RepoSettingsPermissionsPage.tsx:76]: Out of curiosity, what was the <div> needed for?

File: web/src/site-admin/backend.tsx [hunk 5]
```diff
@@ -208,6 +208,40 @@ export function checkMirrorRepositoryConnection(
...
+            mutation ScheduleRepositoryPermissionsSync($repository: ID!) {
+                scheduleRepositoryPermissionsSync(repository: $repository) {
+                    alwaysNil
+                }
+            }
+        `,
+        args
+    ).pipe(
+        map(dataOrThrowErrors),
+        tap(() => resetAllMemoizationCaches()),
```
Comment (eseliger) [inline web/src/site-admin/backend.tsx:223]: @sourcegraph/web this seems to retrigger `every` observable? If so, I think a simple repo refetch would be enough here?

File: web/src/repo/settings/RepoSettingsPermissionsPage.tsx (hunk 4 above)
Comment (unknwon) [reply web/src/repo/settings/RepoSettingsPermissionsPage.tsx:76]: TBH I don't know, I copied the whole thing from

https://github.com/sourcegraph/sourcegraph/blob/68e6e33f8622aecbeeca51599ad7ca14fc368fb4/web/src/repo/settings/RepoSettingsMirrorPage.tsx#L110

File: web/src/user/settings/auth/UserSettingsPermissionsPage.tsx (hunk 1 above)
Comment (unknwon) [reply web/src/user/settings/auth/UserSettingsPermissionsPage.tsx:56]: This whole page isn't viewable by non site-admin, so we don't need to do extra check here.

File: web/src/site-admin/backend.tsx (hunk 5 above)
Comment (unknwon) [reply web/src/site-admin/backend.tsx:223]: I also copied from 
https://github.com/sourcegraph/sourcegraph/blob/68e6e33f8622aecbeeca51599ad7ca14fc368fb4/web/src/site-admin/backend.tsx#L181

Please feel free to make suggestions!

**Comment by unknwon** [comment]
Merging as-is, and happy to address any post-merge comments!

**Comment by christinaforney** [comment]
> cc @christinaforney do you think the screenshot in the description is good enough? (both user and repository ones are look very similar in layout, so just provide one of them is sufficient IMO. Of course I can provide both if you want)

@unknwon - that screenshot looks good, I don't think we need both. Can you use that same page but include the full view of Sourcegraph in the screenshot so it shows where/what we are looking at?

**Comment by unknwon** [comment]
image

@christinaforney is this one better?

**Comment by christinaforney** [comment]
@unknwon - yep that's perfect!

**Pull request merged**